Initial work on documentation generation tool as rewrite step

cgranade opened the pull request:
This PR beings work on a new documentation generation tool that directly outputs DocFX-Flavored Markdown (DFM) from the Q# compilation process, instead of the YAML → template process currently used. This enables:

- Direct usage by third-party libraries (i.e..: no additional template needed).
- Resolving long-standing documentation limitations (e.g.: https://github.com/microsoft/qsharp-compiler/issues/113), and fixing bugs in templates (https://github.com/MicrosoftDocs/quantum-docs-pr/issues/991).
- Adapting output generation for new language features.

The compilation rewrite step in this PR also embeds documentation metadata into new attributes (PR on qsharp-runtime forthcoming), so that downstream tools such as IQ# can consume documentation without re-parsing documentation comments.

By way of implementation, this PR uses the existing documentation output tool as a parser to power the new compilation step.

NB: This PR replaces and supercedes https://github.com/microsoft/qsharp-runtime/pull/253.

Breaking changes:
- UIDs generated for Q# are no longer lower-cased, following standards for UIDs in other languages and frameworks. This will require fixing existing `@""` and `xref:` links to use fully-qualified names as UIDs instead of transforming to lowercase.

TODO:
- [x] Open corresponding PRs on qsharp-runtime, iqsharp, and QuantumKatas with changes needed to support this PR.
- [x] Finish extension method for converting type signatures to Markdown links (operations and tuples are currently `TODO`).
- [ ] Adapt QuantumLibraries documentation for UDTs to include new `/// # Named Items` documentation section.
- [ ] Adapt QuantumLibraries projects to provide package name.
- [x] Adapt symbol encoder in IQ# to utilize new information.
- [x] Package new step as nuget package.
- [x] Adapt Microsoft.Quantum.Sdk to automatically import new package when outputting documentation.
- [x] Generate/update TOC and namespace pages
- [x] Add new MSBuild properties to documentation.

Currently in draft for early feedback, targeting new feature branch.

cgranade commented:
There's still several items remaining to do, such that this is not ready for merging to `master` / `main`, but I think this is now to the point where it's a good place to start the feature branch.

samarsha commented:
Is it possible to merge in updates from master into the feature branch in a separate PR to make the diff here cleaner? It's quite large and looks like it contains unrelated changes from new commits on master.

cgranade commented:
> Is it possible to merge in updates from master into the feature branch in a separate PR to make the diff here cleaner? It's quite large and looks like it contains unrelated changes from new commits on master.

Sounds good, let me go on and do that, then.

cgranade commented:
@samarsha: Please see https://github.com/microsoft/qsharp-compiler/pull/611. Thanks!

samarsha commented on src/QsCompiler/DocumentationGenerator/DocumentationGenerator.csproj:
```diff
@@ -0,0 +1,36 @@
+    </Content>
+  </ItemGroup>
+
+</Project>
```
Can StyleCop be added to this project, like the other C# projects in this repo?

samarsha commented on src/QsCompiler/DocumentationGenerator/DocumentationGenerator.csproj:
```diff
@@ -0,0 +1,36 @@
+  </PropertyGroup>  
+
+  <ItemGroup>
+    <ProjectReference Include="../Compiler/Compiler.csproj" />
```
```suggestion
    <ProjectReference Include="..\Compiler\Compiler.csproj" />
```

samarsha commented on src/QsCompiler/DocumentationGenerator/DocumentationWriter.cs:
```diff
@@ -0,0 +1,221 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
```
Since this is a new project, can nullable be enabled in the csproj instead of per file?

samarsha commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+# We can now get Mono itself.
+RUN curl -o /usr/local/bin/nuget.exe https://dist.nuget.org/win-x86-commandline/latest/nuget.exe && \
+    # Create as alias for nuget
+    echo "alias nuget=\"mono /usr/local/bin/nuget.exe\"" >> /root/.bash_aliases
```
What are the dockerfile and dev container needed for?

cgranade replied:
That was for my own ease of development, as it allows for a reproducible development environment when used in VS Code.

samarsha commented on utilities/summarize_documentation/summarize_documentation.py:
```diff
@@ -0,0 +1,112 @@
+        f.write(warning_comment + yaml.dump(toc_page))
+
+if __name__ == "__main__":
+    main()
```
What is the Python script used for? Instead of doc generator -> Markdown -> YAML, could the doc generator generate both Markdown and YAML directly?

cgranade replied:
That's a bit tricky to do, as the YAML files have to summarize the Markdown files from multiple different assemblies (e.g.: to make a TOC for all Q# APIs in a single reference). I'll add a bit to it, but the purpose of this Python script is to produce those summary YAML files from a collection of different Markdown files. By contrast with our current approach, this makes documentation building stateless, as the documentation generation step no longer needs to read in an existing summary and update it with new content.

cgranade replied:
My thought was to keep using the Markdown parsing parts of DocsParser, but to use the new rewrite steps to generate output instead of the YAML writing functionality in DocsParser. Happy to mark those methods with Obsolete if that would make sense?

bettinaheim replied:
Yes, I think it would be good to mark these things as obsolete and keep just one setup also in the SDK (without the option to configure it to either be markdown or yaml). Then we can eventually remove these methods, giving us one project for the parsing of hte doc comments, one for the emission of the markdown, and a separate tool to convert the markdown to yaml. Does that sound reasonable?

cgranade replied:
Sounds good, I'll remove the QsharpDocsGenerationMode property, then, and will lock everything down to the new Markdown mode. To support that, will mark any YAML output methods in DocsParser as Obsolete.

crazy4pi314 commented:
This is looking pretty great! I know the qRAM team could really make use of this (https://github.com/qsharp-community/qram/issues/27) so that we could take the API docs that we have now and easily convert them to a static site for our users 💖 Let me know if there is anything I can do to help move this along!

bettinaheim commented on src/QuantumSdk/Sdk/Sdk.targets:
```diff
@@ -81,6 +81,8 @@
       <_QscCommandPredefinedAssemblyProperties Condition="$(ExecutionTarget) != ''">$(_QscCommandPredefinedAssemblyProperties) ExecutionTarget:$(ExecutionTarget)</_QscCommandPredefinedAssemblyProperties>
       <_QscCommandPredefinedAssemblyProperties Condition="$(ExposeReferencesViaTestNames)">$(_QscCommandPredefinedAssemblyProperties) ExposeReferencesViaTestNames:true</_QscCommandPredefinedAssemblyProperties>
+      <_QscCommandPredefinedAssemblyProperties Condition="$(QsharpDocsPackageId) != ''">$(_QscCommandPredefinedAssemblyProperties) DocsPackageId:$(QsharpDocsPackageId)</_QscCommandPredefinedAssemblyProperties>
+      <_QscCommandPredefinedAssemblyProperties Condition="$(QsharpDocsGeneration)">$(_QscCommandPredefinedAssemblyProperties) DocsOutputPath:$(QsharpDocsOutputPath)</_QscCommandPredefinedAssemblyProperties>
```
Do you think this is something that other rewrite steps should know about? If there is no reason to add it then I'd rather not. If there is one, then please add the corresponding item to AssemblyConstants in ReservedKeywords.fs.

bettinaheim commented on src/QuantumSdk/DefaultItems/DefaultItems.props.v.template:
```diff
@@ -34,6 +34,9 @@
     <QsharpDocsGeneration>false</QsharpDocsGeneration>
+    <QsharpDocsPackageId>$(PackageId)</QsharpDocsPackageId>
+    <!-- Can be "yaml" for DocParser-based, or "markdown" for RewriteStep-based. -->
+    <QsharpDocGenerationMode>markdown</QsharpDocGenerationMode>
```
Do we really want to permanently maintain both? I'd rather go with one docs generation tool if possible rather than maintain both. 
If we were to keep this, then it needs to be documented in the readme of the Sdk, and ideally we would fail the build or at least emit a warning if this is not set to a recognized mode.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocgenStep.cs:
```diff
@@ -0,0 +1,80 @@
+            return true;
+        }
+
+        public bool PostconditionVerification(QsCompilation compilation) =>
```
I'm ok with leaving it as is, but you could check here that all files the you would expect to have been generated indeed were generated.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocgenStep.cs:
```diff
@@ -0,0 +1,80 @@
+                    Stage = IRewriteStep.Stage.PreconditionVerification
+                });
+
+                foreach (var item in AssemblyConstants)
```
Do you really need to list all assembly constants here? I can see that you would list the ones that you use as part of this rewrite step, but all just seems like unnecessary noise here.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocumentationGenerator.csproj:
```diff
@@ -0,0 +1,43 @@
+    <PackageName>Microsoft.Quantum.DocumentationGenerator</PackageName>
+    <PackageReleaseNotes>See: https://docs.microsoft.com/en-us/quantum/relnotes/</PackageReleaseNotes>
+    <PackageLicenseExpression>MIT</PackageLicenseExpression>
+    <PackageProjectUrl>https://github.com/microsoft/qsharp-runtime</PackageProjectUrl>
```
The url is incorrect here.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocgenStep.cs:
```diff
@@ -0,0 +1,80 @@
+                AssemblyConstants.TryGetValue("DocsOutputPath", out var path)
+                ? path
+                : null,
+                AssemblyConstants.TryGetValue("DocsPackageId", out var packageName)
```
The DocsPackageId doesn't seem to be set anywhere. Do we really need it?

bettinaheim commented on src/QsCompiler/DocumentationGenerator/Extensions.cs:
```diff
@@ -0,0 +1,419 @@
+                    attributeName,
+                    // The following populates all of the metadata needed for a
+                    // Q# literal of type (String, String).
+                    new TypedExpression(
```
Please use SyntaxGenerator.TupleLiteral instead.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/Extensions.cs:
```diff
@@ -0,0 +1,419 @@
+        internal static bool IsDeprecated(this QsCustomType type, out string? replacement) =>
+            type.Attributes.IsDeprecated(out replacement);
+
+        internal static bool IsDeprecated(this IEnumerable<QsDeclarationAttribute> attributes, [NotNullWhen(true)]  out string? replacement)
```
Since the compiler already needs and implements the same logic, please us SymbolResolution.TryFindRedirect instead.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/Extensions.cs:
```diff
@@ -0,0 +1,419 @@
+                )}"
+            };
+
+        internal static string ToSyntax(this QsCallable callable)
```
Aren't all these ToSyntax methods for displaying the Q# representation as it would be in the source code? 
Please use SyntaxTreeToQsharp wherever possible instead.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/ProcessDocComments.cs:
```diff
@@ -0,0 +1,138 @@
+            this.Namespaces = new ProcessDocComments.NamespaceTransformation(this, writer);
+        }
+
+        public QsCompilation OnCompilation(QsCompilation compilation) =>
```
This can be removed; OnCompilation is now defined as part of SyntaxTreeTransformation.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocgenStep.cs:
```diff
@@ -0,0 +1,80 @@
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.Quantum.QsCompiler;
+using Microsoft.Quantum.QsCompiler.Experimental;
```
```suggestion
```

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocgenStep.cs:
```diff
@@ -0,0 +1,80 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
```
```suggestion
```

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocgenStep.cs:
```diff
@@ -0,0 +1,80 @@
+
+namespace Microsoft.Quantum.Documentation
+{
+    public class DocumentationGenerationStep : IRewriteStep
```
Could you please add a minimal doc comment? And since none of the other steps are ending in *Step, it would be make sense to just call it DocumentationGeneration here, too.

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocumentationWriter.cs:
```diff
@@ -0,0 +1,219 @@
+namespace Microsoft.Quantum.Documentation
+{
+
+    public class DocumentationWriter
```
Could you please add documentation to everything here that is part of the public API?

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocumentationWriter.cs:
```diff
@@ -0,0 +1,219 @@
+
+    public class DocumentationWriter
+    {
+        private readonly string outputPath;
```
```suggestion
```

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocumentationWriter.cs:
```diff
@@ -0,0 +1,219 @@
+    public class DocumentationWriter
+    {
+        private readonly string outputPath;
+        public string OutputPath => outputPath;
```
```suggestion
        public string OutputPath { get; }
```

bettinaheim commented on src/QsCompiler/DocumentationGenerator/DocumentationWriter.cs:
```diff
@@ -0,0 +1,219 @@
+using System.IO;
+using System.Threading.Tasks;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler;
```
The following are not needed:

using Microsoft.Quantum.QsCompiler;
using Microsoft.Quantum.QsCompiler.DataTypes;
using Microsoft.Quantum.QsCompiler.SyntaxTokens;

bettinaheim commented on src/QsCompiler/DocumentationGenerator/Extensions.cs:
```diff
@@ -0,0 +1,419 @@
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
```
```suggestion
```

bettinaheim commented on src/QsCompiler/DocumentationGenerator/Extensions.cs:
```diff
@@ -0,0 +1,419 @@
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
```
```suggestion
```

bettinaheim commented on src/QsCompiler/DocumentationGenerator/Extensions.cs:
```diff
@@ -0,0 +1,419 @@
+
+    internal class Callable : IAttributeBuilder<QsCallable>
+    {
+        private QsCallable callable;
```
```suggestion
        private readonly QsCallable callable;
```

bettinaheim commented on src/QsCompiler/DocumentationGenerator/Extensions.cs:
```diff
@@ -0,0 +1,419 @@
+            SyntaxTreeToQsharp.Default.ToCode(type);
+
+        internal static string ToSyntax(this QsTypeItem item) =>
+            item switch
```
A bunch of your switch expressions don't cover all cases (the default usually is missing). Please add that; ideally you could even generate a proper diagnostic for the rewrite step in that case.

bettinaheim reviewed (changes requested):
Thanks for getting us a more robust docs gen tool!
Could you please make a pass over the DocumentationGenerator project and see that it conforms with the style rules defined for the repo? Thanks!

cgranade commented on utilities/summarize_documentation/README.md:
````diff
@@ -0,0 +1,11 @@
+For example:
+
+```bash
+$ pipenv install
````
Ah, this was intended for using Pipfile; I'll modify to point to requirements.txt. Thanks for catching it.

cgranade reviewed (commented):
Thanks for your comments, @bettinaheim! I'll work on addressing those accordingly.

cgranade commented:
/AzurePipelines run

cgranade commented:
@bettinaheim: Thanks again for the review! I've addressed most of your feedback, save for your comments about the `QSharpDocsPackageId` MSBuild property. If it's alright, I'd like to leave that for two reasons:

- When publishing API documentation for a group of packages at once, it may not be clear what package users should depend on to get access to a given function, operation, or UDT. By keeping the package name as metadata, the documentation generation tool can provide this information to users when it is appropriate to do so.
- Some packages may be available under a different package ID than they are built with (e.g.: when metapackages are used to simplify package deployment), such that individual packages should be able to override the property to specify the name of their corresponding metapackage instead.

cgranade commented:
/AzurePipelines run

cgranade commented:
/AzurePipelines run

cgranade commented:
/AzurePipelines run

cgranade commented:
/AzurePipelines run

cgranade commented:
/AzurePipelines run

cgranade commented:
/AzurePipelines run

cgranade commented:
/AzurePipelines run

cgranade commented:
@bettinaheim, @samarsha: Looks like tests pass, now that the CI outage is resolved. Would you be willing to take another look? Thanks!

bettinaheim commented on src/Documentation/DocumentationGenerator/ProcessDocComments.cs:
```diff
@@ -0,0 +1,153 @@
+                          : new DocumentationWriter(outputPath, packageName);
+
+            // We provide our own custom namespace transformation, and expression kind transformation.
+            this.Namespaces = new ProcessDocComments.NamespaceTransformation(this, this.writer);
```
Could you please disable all other transformations? I think since you need the namespace transformation only, you can likely also just disable everything in the base ("base(new TransformationState(), TransformationOptions.Disabled)").

cgranade replied:
Will do, let me apply that fix. Thanks!

bettinaheim commented on src/QuantumSdk/DefaultItems/DefaultItems.props.v.template:
```diff
@@ -34,6 +34,9 @@
     <IncludeQsharpCorePackages>true</IncludeQsharpCorePackages> 
     <IncludeProviderPackages>true</IncludeProviderPackages> 
     <QsharpDocsGeneration>false</QsharpDocsGeneration>
+    <QsharpDocsPackageId>$(PackageId)</QsharpDocsPackageId>
```
Ah, I misunderstood what this is. Could you please document that property in the readme of the Sdk then?

bettinaheim commented on src/QuantumSdk/README.md:
```diff
@@ -88,111 +88,113 @@ The content of the file should be similar to the following, with `Package_Name`
 ## Known Issues ##
 
 The following issues and PRs may be of interest when using the Sdk:
-> https://github.com/NuGet/Home/issues/8692    
```
I am not sure if a formatter somehow did the whitespace removal, but could you please keep all whitespace in the markdown file as is, since it is needed...

cgranade replied:
Ah, thanks for the catch! A formatter must have done so, but I'll revert that change.

cgranade commented:
/azp run
